addOns/jwt/src/main/java/org/zaproxy/zap/extension/jwt/utils/JWTUtils.java

```diff
@@ -0,0 +1,248 @@
+     * (for finding vulnerabilities in web applications that are using old implementations or custom
+     * implementations) is not possible therefore added this custom implementation for HS*
+     * algorithms.
+     *
+     * <p>
+     *
+     * @param token
+     * @param secretKey
+     * @param algorithm
+     * @return
+     * @throws JWTException
+     */
+    public static String getBase64EncodedHMACSignedToken(
+            byte[] token, byte[] secretKey, String algorithm) throws JWTException {
+        try {
+            if (JWTConstants.JWT_HMAC_ALGO_TO_JAVA_ALGORITHM_MAPPING.containsKey(algorithm)) {
+                Mac hmacSHA =
+                        Mac.getInstance(
+                                JWTConstants.JWT_HMAC_ALGO_TO_JAVA_ALGORITHM_MAPPING.get(
+                                        algorithm));
+                SecretKeySpec hmacSecretKey = new SecretKeySpec(secretKey, hmacSHA.getAlgorithm());
+                hmacSHA.init(hmacSecretKey);
+                byte[] tokenSignature = hmacSHA.doFinal(token);
+                String base64EncodedSignature =
+                        JWTUtils.getBase64UrlSafeWithoutPaddingEncodedString(tokenSignature);
+                return base64EncodedSignature;
+            } else {
+                throw new JWTException(algorithm + " is not a supported HMAC algorithm.");
+            }
+        } catch (InvalidKeyException | NoSuchAlgorithmException e) {
+            throw new JWTException(
+                    "Exception occurred while Signing token: " + getString(token), e);
+        }
+    }
+
+    private static boolean hasBearerToken(String value) {
+        return Pattern.compile(BEARER_TOKEN_REGEX).matcher(value).find();
+    }
+
+    /**
+     * This utility method removes {@literal BEARER_TOKEN_REGEX} from the value. For now it is just
+     * removing {@literal BEARER_TOKEN_REGEX} but in future we might need to remove other type of
+     * schemes too.
+     *
+     * @param value
+     * @return
+     */
+    public static String extractingJWTFromParamValue(String value) {
+        if (hasBearerToken(value)) {
+            value = value.replaceAll(BEARER_TOKEN_REGEX, "").trim();
+        }
+        return value;
+    }
+
+    /**
+     * This utility method adds the {@literal BEARER_TOKEN_KEY} to the value. This method reverses
+     * the operation performed by {@link JWTUtils#extractingJWTFromParamValue}
+     *
+     * @param value
+     * @param jwtToken
+     * @return
+     */
+    public static String addingJWTToParamValue(String value, String jwtToken) {
+        if (hasBearerToken(value)) {
+            jwtToken = BEARER_TOKEN_KEY + " " + jwtToken;
+        }
+        return jwtToken;
+    }
+
+    /**
+     * @param customFieldFuzzer
+     * @param clonedJWTokenBean
+     * @throws JWTException
+     */
+    public static void handleSigningOfTokenCustomFieldFuzzer(
```

Comment:
is there any parser for Private Key based on RSA present in ZAP, i think it should be there as we are giving a way to upload and download certificates. @kingthorin and @thc202  please point me to that file so that i can parse the key String and work on that as the way i have done is quite hacky.

Reply:
There is pkcs12 and pkcs11 handling in the core. Will that help?

Check src/org/parosproxy/paros/extension/option/OptionsCertificatePanel.java and related pieces.